lib/virtio/virtqueue.c

```diff
@@ -287,6 +287,41 @@ void *virtqueue_get_available_buffer(struct virtqueue *vq, uint16_t *avail_idx,
 	return buffer;
 }
 
+unsigned int virtqueue_buffer_writable(struct virtqueue *vq, uint16_t avail_idx)
+{
+	uint16_t flags;
+
+	VRING_INVALIDATE(vq->vq_ring.desc[avail_idx]);
+	flags = vq->vq_ring.desc[avail_idx].flags;
+
+	return !!(flags & VRING_DESC_F_WRITE);
+}
+
+void *virtqueue_get_next_buffer(struct virtqueue *vq, uint16_t *avail_idx,
+				uint32_t *len)
+{
+	uint16_t flags;
+	void *buffer;
+
+	VRING_INVALIDATE(vq->vq_ring.desc[*avail_idx]);
+	flags = vq->vq_ring.desc[*avail_idx].flags;
```

Comment:
As we access to the vring here i would prefer to to keep `VQUEUE_BUSY` protection... you can use a goto if you want to have a single call to  `VQUEUE_IDLE()` in  he function